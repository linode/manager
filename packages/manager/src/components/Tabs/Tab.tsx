import { Tab as ReachTab } from '@reach/tabs';
import * as React from 'react';
import { makeStyles } from 'tss-react/mui';

import type { Theme } from '@mui/material/styles';
import type { TabProps } from '@reach/tabs';

const useStyles = makeStyles()((theme: Theme) => ({
  tab: {
    '&[data-reach-tab]': {
      '&:focus': {
        backgroundColor: theme.color.grey7,
      },
      '&:hover': {
        backgroundColor: theme.color.grey7,
      },
      alignItems: 'center',
      borderBottom: '2px solid transparent',
      color: theme.textColors.linkActiveLight,
      display: 'inline-flex',
      flexShrink: 0,
      fontSize: '0.9rem',
      lineHeight: 1.3,
      marginTop: theme.spacing(0.5),
      maxWidth: 264,
      minHeight: theme.spacing(5),
      minWidth: 50,
      padding: '6px 16px',
      textDecoration: 'none',
    },
    '&[data-reach-tab][data-selected]': {
      borderBottom: `3px solid ${theme.textColors.linkActiveLight}`,
      color: theme.textColors.headlineStatic,
      font: theme.font.bold,
    },
  },
}));

interface TabPropsWithClassName extends TabProps {
  className?: string;
}

const Tab = ({ children, className, ...rest }: TabPropsWithClassName) => {
  const { classes, cx } = useStyles();

  return (
    <ReachTab className={cx(classes.tab, className)} {...rest}>
      {children}
    </ReachTab>
  );
};

export { Tab };
