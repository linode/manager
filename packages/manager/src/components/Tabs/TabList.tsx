import { styled } from '@mui/material/styles';
import { TabList as ReachTabList } from '@reach/tabs';
import * as React from 'react';

import type { TabListProps } from '@reach/tabs';

interface TabListPropsWithClassName extends TabListProps {
  className?: string;
}

const TabList = ({
  children,
  className,
  ...rest
}: TabListPropsWithClassName) => {
  return (
    <StyledReachTabList className={className} {...rest}>
      {children}
    </StyledReachTabList>
  );
};

export { TabList };

const StyledReachTabList = styled(ReachTabList)(({ theme }) => ({
  '&[data-reach-tab-list]': {
    background: 'none !important',
    boxShadow: `inset 0 -1px 0 ${theme.borderColors.divider}`,
    marginBottom: theme.spacingFunction(16),
    [theme.breakpoints.down('lg')]: {
      overflowX: 'auto',
      padding: 1,
    },
  },
  color: theme.color.tableHeaderText,
}));
