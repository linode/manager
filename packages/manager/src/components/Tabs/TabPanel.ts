export { TabPanel } from '@reach/tabs';
