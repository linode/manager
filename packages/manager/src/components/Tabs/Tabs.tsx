export { Tabs } from '@reach/tabs';
