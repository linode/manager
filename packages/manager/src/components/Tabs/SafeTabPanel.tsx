import { useTabsContext } from '@reach/tabs';
import * as React from 'react';

import { TabPanel } from './TabPanel';

interface SafeTabPanelProps {
  children: React.ReactNode | null;
  index: null | number;
}

/**
 * SafeTabPanel only renders its children when it is currently selected. This is helpful when a
 * child component of SafeTabPanel requests data when it is mounted. In this case, we may want to
 * avoid mounting the component until the tab is actually selected/visible.
 */
export const SafeTabPanel = (props: SafeTabPanelProps) => {
  const { index, ...tabPanelProps } = props;

  const { selectedIndex } = useTabsContext();
  const isSelected = selectedIndex === index;

  return (
    <TabPanel {...tabPanelProps}>{isSelected ? props.children : null}</TabPanel>
  );
};
