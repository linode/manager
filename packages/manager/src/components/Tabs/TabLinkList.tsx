import * as React from 'react';
import { Link } from 'react-router-dom';

import { Tab } from 'src/components/Tabs/Tab';
import { TabList } from 'src/components/Tabs/TabList';

export interface Tab {
  routeName: string;
  title: string;
}

interface TabLinkListProps {
  noLink?: boolean; // @todo: remove this prop if we use NavTab widely.
  tabs: Tab[];
}

export const TabLinkList = ({ noLink, tabs }: TabLinkListProps) => {
  return (
    <TabList>
      {tabs.map((tab, _index) => {
        // @todo: remove this if we use NavTab widely.
        const extraTemporaryProps: any = noLink
          ? {}
          : { as: Link, to: tab.routeName };

        return (
          <Tab
            key={`tab-${_index}`}
            // @todo: remove this if we use NavTab widely.
            {...extraTemporaryProps}
          >
            {tab.title}
          </Tab>
        );
      })}
    </TabList>
  );
};
