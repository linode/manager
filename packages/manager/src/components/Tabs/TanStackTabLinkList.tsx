import { Link as TanstackLink } from '@tanstack/react-router';
import * as React from 'react';

import { Tab } from 'src/components/Tabs/Tab';
import { TabList } from 'src/components/Tabs/TabList';

import type { Tab as TanstackTab } from 'src/hooks/useTabs';

export interface Tab {
  chip?: React.JSX.Element | null;
  routeName: string;
  title: string;
}

interface TabLinkListProps {
  noLink?: boolean;
  tabs: TanstackTab[];
}

export const TanStackTabLinkList = ({ noLink, tabs }: TabLinkListProps) => {
  return (
    <TabList>
      {tabs.map((tab, _index) => {
        return (
          <Tab
            // @ts-expect-error - Tab accepts 'as' prop at runtime but it's not in the types
            as={noLink ? undefined : TanstackLink}
            key={`tab-${_index}`}
            preload={noLink ? undefined : 'intent'}
            to={noLink ? undefined : tab.to}
          >
            {tab.title}
            {tab.chip}
          </Tab>
        );
      })}
    </TabList>
  );
};
