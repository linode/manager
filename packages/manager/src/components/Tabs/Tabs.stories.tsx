import * as React from 'react';

import { Paper } from 'src/components/Paper';
import { SafeTabPanel } from 'src/components/Tabs/SafeTabPanel';
import { TabPanels } from 'src/components/Tabs/TabPanels';
import { Tabs } from 'src/components/Tabs/Tabs';
import { Typography } from 'src/components/Typography';

import { TabLinkList } from './TabLinkList';

import type { TabsProps } from '@reach/tabs';
import type { Meta, StoryObj } from '@storybook/react';

const _tabs = [
  {
    routeName: '/tab1',
    title: 'Tab 1',
  },
  {
    routeName: '/tab2',
    title: 'Tab 2',
  },
  {
    routeName: '/tab3',
    title: 'Tab 3',
  },
];

/**
 * Tabs allow users to view content across categories that are all related and share a similar hierarchy.
 * - Each tab should contain distinct content
 * - Tabs can scroll horizontally on smaller screens and mobile
 * - Content across a tab set should be considered peers, each tab being of equal importance
 */
export const Default: StoryObj<TabsProps> = {
  render: () => (
    <Tabs>
      <TabLinkList tabs={_tabs} />
      <TabPanels>
        {_tabs.map((tab, idx) => (
          <SafeTabPanel index={idx} key={`tab-${idx}`}>
            <Paper sx={{ padding: 2 }}>
              <Typography variant="body1">Content for {tab.title}</Typography>
            </Paper>
          </SafeTabPanel>
        ))}
      </TabPanels>
    </Tabs>
  ),
};

const meta: Meta<TabsProps> = {
  argTypes: {
    defaultIndex: {
      control: { type: 'number' },
      description: 'Starts the tabs at a specific index.',
    },
    index: {
      control: { type: 'number' },
      description:
        'The index of the tab, must include `onChange` in order for the tabs to be interactive.',
    },
    onChange: {
      control: false,
      description: 'Callback fired when the value changes.',
    },
  },
  args: {
    defaultIndex: 0,
    index: 0,
    onChange: () => null,
  },
  component: Tabs,
  title: 'Components/Tabs',
};
export default meta;
