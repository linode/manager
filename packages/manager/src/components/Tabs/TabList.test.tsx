import { fireEvent, render } from '@testing-library/react';
import React from 'react';

import { wrapWithTheme } from 'src/utilities/testHelpers';

import { Tab } from './Tab'; // Adjust the import path based on your project structure
import { TabList } from './TabList';
import { Tabs } from './Tabs'; // Adjust the import path based on your project structure

describe('TabList component', () => {
  it('renders TabList correctly', () => {
    const { container } = render(
      wrapWithTheme(
        <Tabs>
          <TabList />
        </Tabs>,
        {
          MemoryRouter: {
            initialEntries: [{ pathname: '/tab-1' }],
          },
        }
      )
    );

    expect(container).toMatchSnapshot();
  });

  it('handles className prop', () => {
    const { container } = render(
      wrapWithTheme(
        <Tabs>
          <TabList className="custom-class" />
        </Tabs>,
        {
          MemoryRouter: {
            initialEntries: [{ pathname: '/tab-1' }],
          },
        }
      )
    );

    const tabListElement = container.querySelector('.custom-class');
    expect(tabListElement).toBeInTheDocument();
  });

  it('handles click events on tabs', () => {
    const { getByText } = render(
      wrapWithTheme(
        <Tabs>
          <TabList>
            <Tab>Tab 1</Tab>
            <Tab>Tab 2</Tab>
          </TabList>
        </Tabs>,
        {
          MemoryRouter: {
            initialEntries: [{ pathname: '/tab-1' }],
          },
        }
      )
    );

    const tab1 = getByText('Tab 1');
    fireEvent.click(tab1);

    expect(tab1).toHaveAttribute('aria-selected', 'true');
  });
});
