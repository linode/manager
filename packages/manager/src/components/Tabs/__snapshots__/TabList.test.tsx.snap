// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TabList component > renders TabList correctly 1`] = `
<div>
  <div
    data-orientation="horizontal"
    data-reach-tabs=""
  >
    <div
      aria-orientation="horizontal"
      class="css-bjm66n"
      data-reach-tab-list=""
      role="tablist"
    />
  </div>
</div>
`;
